### PR TITLE
rewrite bundling for smaller bundles

### PR DESCRIPTION
**Pull request by devsnek**
- Remove redundant async functions in common cases
- Replace validation with smaller call to WebAssembly.validate
- Remove duplicated check for shared memory
- Removes `fetch(data uri)` to avoid CSP.

Overall, this results in ~6% size savings (621 bytes vs 587 bytes

### CODE REVIEW

**Comment by surma** [comment]
So I like where this is going, thanks for your effort!

Two things:

1. The `index.js` files contained `// Name:` and `// Proposal:` comments that are used to programmatically update the readme. This would need to be moved to the `.wat` file.
2. This is actually _bigger_ after gzip (621 bytes on master vs 686 bytes on this branch). I think you might be better off _not_ trimming the wasm preamble as gzip will take care of that.

An idea I’ve had (and maybe you wanna take a stab at that): I wondered if hex encoding is better. 
- The decoder is literally `new Uint8Array(hexstring.split(/(..)/).filter(x => x).map(v => parseInt(v, 16)))`
- Same code works in node and web
- It probably compresses just as well as base64 with gzip

WDYT?

**Comment by devsnek** [comment]
@surma what are you using to test the gzipped size. by my count we are now saving exactly 1 byte, *but* we also no longer rely on fetch.

**Comment by surma** [comment]
`cat dist/esm/index.js | gzip -c9n | wc -c`

**Comment by devsnek** [comment]
in that case we're now down to 587

**Comment by surma** [comment]
Brilliant! Thanks a bunch!

I’m on mobile for today, but I’ll give this a proper review tomorrow 🎉

File: rollup-plugins/size-printer.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { gzipSync } from "zlib";
+
+export default function sizePrinter() {
```
Comment (surma) [inline rollup-plugins/size-printer.js:3]: Good thinking on this one :D

**Review by surma — APPROVED** [review]

**Comment by surma** [comment]
I removed the global `data` object as it would make the code not as tree-shakable. We are also down to 560 Bytes :D

If you have on objections, I’ll merge this :)

**Comment by surma** [comment]
@googlebot I consent.

**Comment by surma** [comment]
... lol

**Comment by devsnek** [comment]
really nice thinking on the `bind(null, hex string)`!

**Pull request merged**